### ISSUE TITLE
Unit freeze glitch detection

### ISSUE THREAD

**Issue by heinermann** [bwapi/bwapi#566]
BWAPI should detect when units get stuck due to the glitch in Starcraft that causes units to freeze.

**Comment by krasi0** [comment]
There already exists a method isStuck for that purpose

**Comment by heinermann** [comment]
No, isStuck checks for units that are on illegal positions on top of other units.

**Comment by chriscoxe** [comment]
I am seeing this happen in quite a large proportion of games, for the occasional zergling.  Do you know if there's a way to detect it is happening (in BWAPI v4.1.2 anyway), e.g. any particular unit flags that are always true in this situation (maybe isAttackFrame())?  Do you know if there is anything special I need to do to enable it to respond to commands after I've detected it (e.g. call Unit::stop())?  Thanks

**Comment by heinermann** [comment]
If you've detected it, ordering the unit to stop should be sufficient to fix it. As far as detecting it is concerned, the BWMemoryEdit tool (written in C#) was created for finding these kinds of flags in the CUnit, CSprite, and CImage structures. It is currently a bit difficult to use, so feel free to improve it. When conditions are discovered, we can make it a boolean field in the BWAPI::Unit structure.

So after this research has been done, you could do something like
~~~cpp
if (myUnit->isGlitched())
  myUnit->stop();
~~~

**Comment by krasi0** [comment]
Why shouldn't isStuck cover this case, too?

**Comment by neivv** [comment]
Here's information on at least one way to make units using iscript.bin movement freeze:

When an unit is moving and acquires an attack target without stopping, it starts playing its attack animation and stops moving one frame afterwards.
However, if the target has moved outside of unit's range excatly one frame afterwards, the unit becomes stuck as it is tries to chase the target, and as it is still "moving" (`movementFlags & 0x2 != 0`), bw will not switch its iscript animation back to movement animation, which contains the `move` commands.
So the unit appears to be stuck.

As such, the following checklist could *maybe* be used to check for this state:
- Unit uses iscript.bin movement, so `unit->flingyMovementType == 2`
- Unit is still "moving" `unit->movementFlags & 0x2 != 0`
- Unit has a target
- Unit's iscript animation is one of the attack animations.
- Unit's target is outside of its attack range

This has not been tested and may have false positives.
